stm32 enables  HSI48 fixed clock with device tree

FRASTM opened the pull request:
Until now the HSI 48 clock of the stm32 mcu is enabled by the peripheral that needs it with LL_RCC_HSI48_Enable
With this PR, it is defined by the DeviceTree for each stm32 that needs it.
As the frequency is fixed at 48MHz, it is not required by the DTS to specify the value.

New sequence added in the stm32_clock_configuration/stm32_common_core/  testcase to check that HSI48 is ready when required.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49870

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: clock_control: some stm32 have a HSI48 fixed clock

For the stm32 devices that have a HSI48 clock,
the driver enables it, like any other fixed clock,
if needed and supported by the serie.
For stm32L0, SYSCFG VREFINT is also required.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
For the STM32F0 serie : F04x, F07x, F09x devices only have the HSI48.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -633,6 +633,23 @@ static void set_up_fixed_clock_sources(void)
 		z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
 	}
+
+#if defined(RCC_HSI48_SUPPORT)
```
```suggestion
	if (IS_ENABLED(STM32_HSI48_ENABLED)) {
```

FRASTM replied:
Only stm32 devices  with STM32_HSI48_ENABLED  defined have the LL_RCC_HSI48_Enable in their LL drivers.
For example on stm32wl55, which has no RCC_HSI48_SUPPORT, has no LL_RCC_HSI48_Enable function

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -633,6 +633,23 @@ static void set_up_fixed_clock_sources(void)
 		z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
 	}
+
+#if defined(RCC_HSI48_SUPPORT)
+#if defined(CONFIG_SOC_SERIES_STM32L0X)
```
```suggestion
	if (IS_ENABLED(STM32_HSI48_ENABLED)) {
#if defined(CONFIG_SOC_SERIES_STM32L0X)
```

FRASTM replied:
this is not totally equivalent : 
- RCC_HSI48_SUPPORT if the HSI48 feature exists in the mcu
- STM32_HSI48_ENABLED if the DTS enables the existing HSI48

erwango replied:
To be moved inside below `if (IS_ENABLED(STM32_HSI48_ENABLED)) {` condition

FRASTM replied:
that could be 
```
#if defined(STM32_HSI48_ENABLED)
	/* For all series with HSI 48 clock support */
	if (IS_ENABLED(STM32_HSI48_ENABLED)) {
```

when `#define STM32_HSI48_ENABLED	1`  only when the node is enabled in the DTS (assuming the HSI48 exists for that device)

FRASTM committed:
boards: arm: nucleo_wb55rg enables hsi48 for Bluetooth

The nucleo_wb55rg platform has Bluetooth feature that
requires the HSI48.
The board DTS enables the HSI48MHz clock for that purpose.
And also selects it as the source (default) of the CLK48.
Note that the clk48 node must be enabled for any another source.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -603,22 +603,12 @@ static int entropy_stm32_rng_init(const struct device *dev)
 #elif defined(RCC_CR2_HSI48ON) || defined(RCC_CR_HSI48ON) \
 	|| defined(RCC_CRRCR_HSI48ON)
```
Is this still required ?

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -633,6 +633,23 @@ static void set_up_fixed_clock_sources(void)
 		z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
 	}
+
+#if defined(RCC_HSI48_SUPPORT)
```
Are we sure all series with HSI48 are defining this switch ?

FRASTM replied:
better use` #if defined(RCC_CR2_HSI48ON) || defined(RCC_CR_HSI48ON) || defined(RCC_CRRCR_HSI48ON)`

erwango commented on dts/bindings/clock/st,stm32-hsi48-clock.yaml:
```diff
@@ -0,0 +1,23 @@
+        Fixed 48MHz output clock frequency,
+        No need to specify it in the DTS
+      required: false
+      default: 48000000
```
Don't use a default value. Please follow what was done for other clocks.
Then this new binding is not required any more and we can just use `fixed-clock` compatible

FRASTM replied:
using fixed_clock instead with `clock-frequency = <DT_FREQ_M(48)>;`

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -420,11 +420,10 @@ static void start_ble_rf(void)
-	while (!LL_RCC_HSI48_IsReady()) {
-	}
-
+	/* HSI48 Clock is enabled through the DTS */
+#if !STM32_HSI48_ENABLED
+#warning Bluetooth IPM requires HSI48 clock to be enabled
+#endif
```
Warning is good, but it should not prevent correct behavior. This code section shouldn't be purely removed, only deprecated.
Apply everywhere.

FRASTM replied:
Keeping HSI48 clock enable and adding a warning in case this is not defined by the DTS

FRASTM committed:
dts: arm: stm32 devices with hsi48 clock

Add the HSI48 clock to the stm32 devices that have this
clock signal.
Within a stm32 family, only few might have the
RCC_HSI48_SUPPORT.
STM32WB has a CLK48 mux.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: stm32: do not enable the HSI48 locally

The HSI48 is enabled by clock control driver.
It is no more done by each driver that requires this clock
However when using rng or sdmmc or bluetooth/ipm or usb,
the HSI48 clock must be present in the DTS.
Add a warning for this particular check but keep the deprecated
HSI48 clock enable : keeping for legacy but to remove later.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -421,11 +421,15 @@ static void start_ble_rf(void)
 
-	/* Set RNG on HSI48 */
+#if !STM32_HSI48_ENABLED
+#warning Bluetooth IPM requires HSI48 clock to be enabled
+	/* Deprecated: Set BLE on HSI48 */
```
We should clarify that HSI48 should be enabled **using device tree**

FRASTM replied:
`#warning Bluetooth IPM requires HSI48 clock to be enabled using device `

erwango commented on drivers/disk/sdmmc_stm32.c:
```diff
@@ -154,13 +154,20 @@ static int stm32_sdmmc_clock_enable(struct stm32_sdmmc_priv *priv)
 	defined(CONFIG_SOC_SERIES_STM32U5X)
+#if !STM32_HSI48_ENABLED
+	/* Deprecated: Set HSI48 */
+#warning USB device requires HSI48 clock to be enabled
```
Ditto for message clarification

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -421,11 +421,15 @@ static void start_ble_rf(void)
 
+#endif /* !STM32_HSI48_ENABLED */
+	/* HSI48 Clock is enabled through the DTS */
 	LL_RCC_SetCLK48ClockSource(LL_RCC_CLK48_CLKSOURCE_HSI48);
```
Defining CLK48 clock source should actually be done using clock_control driver using a [clock mux](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/clock_control/clock_stm32_mux.c).

See [h7 example](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/clock_control/stm32_clock_configuration/stm32h7_devices/boards/spi1_per_ck_hsi.overlay).

So:
- this call should be removed.
- In theory, we should define a clk_48 clock mux node that selects HSI48 as source (could be default in stm32wb.dtsi). Though since here we only default value, this would be adding a (simple) driver for nothing. So this could be left as disabled in stm32wb.dtsi file
- on nucleo_wb, add the node with `clocks = <&rcc STM32_SRC_HSI48 CLK48_SEL(0)>;` as example. But keep it disabled to save some footprint (since this is the default value)

erwango commented on dts/arm/st/f0/stm32f072.dtsi:
```diff
@@ -8,6 +8,15 @@
 
 / {
 	soc {
+		clocks {
+			clk_hsi48: clk-hsi48 {
+				#clock-cells = <0>;
+				compatible = "fixed-clock";
+				clock-frequency = <DT_FREQ_M(48)>;
+				status = "disabled";
+			};
+		};
+
```
Not required. F072 inlcudes F070

FRASTM replied:
removed from stm32f072.dtsi

erwango commented on include/zephyr/dt-bindings/clock/stm32wb_clock.h:
```diff
@@ -22,7 +22,7 @@
 /** Fixed clocks  */
 #define STM32_SRC_HSI		0x001
-/* #define STM32_SRC_HSI48	0x002 */
+#define STM32_SRC_HSI48	0x002
```
```suggestion
#define STM32_SRC_HSI48		0x002
```

FRASTM replied:
done

erwango commented on boards/arm/nucleo_wb55rg/nucleo_wb55rg.dts:
```diff
@@ -97,6 +97,15 @@
+};
+
+&clk48 {
+	/* default source is HSI48 */
+	clocks = <&rcc STM32_SRC_HSI48 CLK48_SEL(0)>;
```
```suggestion
	/* Node is disabled by default as default source is HSI48 */
	/* To select another clock, enable the node */
	clocks = <&rcc STM32_SRC_HSI48 CLK48_SEL(0)>;
```

FRASTM replied:
comment is changed

ABOSTM commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -613,13 +616,18 @@ static int entropy_stm32_rng_init(const struct device *dev)
 	LL_SYSCFG_VREFINT_EnableHSI48();
 #endif /* CONFIG_SOC_SERIES_STM32L0X */
 
+#if !defined(CONFIG_SOC_SERIES_STM32WBX)
```
The use of HSEM has been especially introduced for WB, to protect against dual core concurrent access. So it looks to me like a non sense to put this hsem under compilation switch !WB

FRASTM replied:
yes, 
Keep z_stm32_hsem_lock in the drivers/entropy/entropy_stm32.c

FRASTM commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -633,6 +633,24 @@ static void set_up_fixed_clock_sources(void)
+	/* For all series with HSI 48 clock support */
+	if (IS_ENABLED(STM32_HSI48_ENABLED)) {
+#if defined(CONFIG_SOC_SERIES_STM32L0X)
+		/*
```
The CFG_HW_CLK48_CONFIG_SEMID is already locked/unlocked by the RNG or USB driver in that particular case when required.

FRASTM commented:
Add comment : the HSI48 enable part is kept for legacy but should be remove later and be configured exclusively by the DTS

FRASTM commented:
rebase to re-run CI

ABOSTM reviewed (dismissed):
lgtm

FRASTM committed:
boards: arm: enabling hsi48 for RNG peripheral on stm32 platforms

The different boards with stm32 which have node enabled in their DTS
also requires the HSI48 clock to be enabled.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: enabling hsi48 for SDMMC peripheral on stm32 platforms

The different boards with stm32 which have node enabled in their DTS
also requires the HSI48 clock to be enabled.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: enabling hsi48 for USB peripheral on stm32 platforms

The different boards with stm32 which have node enabled in their DTS
also requires the HSI48 clock to be enabled.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Enabling &clk_hsi48 node for all the boards with stm32 RNG or USB or SDMMC or BLE peripheral enabled.

FRASTM committed:
include: bindings: stm32 clock HSI48 defined

The stm32 devices which have a HSI48 clock possibility.
for stm32F0, stm32G0, stm32G4, stm32H7, stm32L0,
stm32L4, stm32L5, stm32U5, stm32WB

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on include/zephyr/dt-bindings/clock/stm32wb_clock.h:
```diff
@@ -22,7 +22,7 @@
 /** Fixed clocks  */
 #define STM32_SRC_HSI		0x001
-/* #define STM32_SRC_HSI48	0x002 */
+#define STM32_SRC_HSI48		0x002
```
Now done in commit 31c1f55c72466040da8217696935adc24ff81a8c

erwango commented on include/zephyr/dt-bindings/clock/stm32h7_clock.h:
```diff
@@ -26,7 +26,7 @@
 #define STM32_SRC_LSE		0x00B
 #define STM32_SRC_LSI		0x00C
 /** Oscillators not yet supported */
```
Let's remove this confusing comment

erwango commented on include/zephyr/dt-bindings/clock/stm32f0_clock.h:
```diff
@@ -20,7 +20,7 @@
 #define STM32_SRC_HSI		0x001
 #define STM32_SRC_LSE		0x002
 #define STM32_SRC_LSI		0x007
-/* #define STM32_SRC_HSI48	0x003 */
+#define STM32_SRC_HSI48		0x003
 /** System clock */
```
Can you fix numbering here also while you're at ?

erwango commented on include/zephyr/dt-bindings/clock/stm32g0_clock.h:
```diff
@@ -20,18 +20,19 @@
 /** PLL clock outputs */
-#define STM32_SRC_PLL_P		0x008
-#define STM32_SRC_PLL_Q		0x009
-#define STM32_SRC_PLL_R		0x00a
+#define STM32_SRC_PLL_P	0x009
+#define STM32_SRC_PLL_Q	0x00a
+#define STM32_SRC_PLL_R	0x00b
```
Fix indentation

erwango reviewed (changes requested):
Otherwise LGTM

FRASTM commented:
fix duplicate case value ` STM32_SRC_PLL_P:`
